executor/index_lookup_join.go

```diff
@@ -613,10 +613,12 @@ func compareRow(sc *stmtctx.StatementContext, left, right []types.Datum) int {
 
 func (iw *innerWorker) fetchInnerResults(ctx context.Context, task *lookUpJoinTask, lookUpContent []*indexJoinLookUpContent) error {
 	innerExec, err := iw.readerBuilder.buildExecutorForIndexJoin(ctx, lookUpContent, iw.indexRanges, iw.keyOff2IdxOff, iw.nextColCompareFilters)
+	if innerExec != nil {
+		defer terror.Call(innerExec.Close)
+	}
```

Comment:
May we avoid this code pattern?


A intuitive code pattern for go programmer is like
```go
innerExec, err := call()
if err != nil { return err } // we assume that a is nil if err happens, and need not call Close
defer innerExec.Close()
```

Could we do some effect on `buildExecutorForIndexJoin` to make sure no need to Close `innerExec` when `err` is not `nil`?

cc @XuHuaiyu